### ISSUE TITLE
Records disappeared from tables catalog_product_link and catalog_product_link_attribute_int

### ISSUE THREAD

**Issue by ADDISON74** [OpenMage/magento-lts#1570]
I faced many times this situation but never investigated. The **catalog_product_link** and **catalog_product_link_attribute_int** tables are empty of data, but auto_increment remains the same. I can easily restore the tables if I have a daily backup, but if I don't find out in time and in the meantime data is added to an emptied table before, I have to look for the last complete version from the past and then add to it everything that exists after. Checking the data consumes me time because there are tens of thousands of records.

At the low level the stores are managed by me, there are users who add products only, but who have not activated Backend sections and no access to the server. Although I constantly analyzed the logs of active services, PHP, Magento, I did not find any errors. Recently I was able to identify the date when happened and I associated possible causes of the issue based on the daily tasks.

The first cause even the users do not have access to anything else other then Products section in Backend, they are making duplicates of the products frequently. It is possible a product that already has Related / Upsell / Cross-Sell attached by duplication to create the issue. Trying to insert or getting duplicate data in the table is becomes corrupted. I have to check this.

The second cause is the import process I did on that day. Although the last imported file on the day of the issue occurred was limited to two columns SKU and Price. I did this task before many times and nothing happened with Related Products. I have to check this too.

It could be a problem in Magento code and I would like to investigate. Does anyone know where I could enter some code to write in a log file whenever changes are made to these tables? Any other advice is welcome.

**Comment by colinmollenhour** [comment]
> Trying to insert or getting duplicate data in the table is becomes corrupted.

Do you mean at the record level or at the database engine level? What database server are you using and have you confirmed there is no bug with the database engine? Maybe the table files are corrupt and just need to be dropped and recreated?

You could also check the foreign keys `ON DELETE` property and change any `CASCADE` to `RESTRICT` in case the rows are being deleted due to an `ON DELETE CASCADE` from another table. Then you will get an error somewhere else when the rows would be deleted and that might help you locate the source.

I've not used the links a lot but I've never seen this happen so I'd be a bit surprised if it was a bug in the code, but I haven't the time to look at all the code related to these tables. Is there any code that deletes rows from these tables by anything other than primary key?

**Comment by ADDISON74** [comment]
It is on the record level. I was using initially MariaDB then I switched to Percona. The same thing happened with both cases. Databases were dropped and recreated from dumps.

The only tables which were affected are those two mentioned in the initial post. All the other tables are just fine. By truncating the tables and importing the records based on a CSV file things were back to normal.

I will follow your advice and see what's happening.At this point I was faced with the effect but I have to identify the cause. I'll look over the code and hope to find something.

On Stack Exchange there are two guys who complains about "disappearing data for Related Product" in frontend. My personal opinion is that it is related to Magento. If there was a problem with the database server I would have had error messages. This situation could be generated by duplicate or import, the only ones actions that were made by users.

Parts of magento that are not commonly used can generate issues. For example, I reported that in the case of reviews, bad things happen in the database tables. It is useless to delete the data in the Backend because orphan information remains in the tables.

**Comment by ADDISON74** [comment]
I will post here my observation related to this issue after testing different scenarios.

**Comment by ADDISON74** [comment]
**CATALOG_PRODUCT_LINK_ATTRIBUTE_INT table - AUTO_INCREMENT issue**

Both **catalog_product_link** and **catalog_product_link_attribute_int** tables have auto_increment values set to 1. There are 5 simple products created before. 

1. I am creating a new simple product then add for Related Products: 1, 2, 3 and click [Save and Continue Edit]. Now in database both auto_increment values are set to 4 which is correct.

2. Add for Up-sells Product: 4, 5 and click [Save and Continue Edit]. A strange behavior occurs in table catalog_product_link_attribute_int. See bellow:

On **value_id** column there is a break in numbering. 7 and 8 should be 4 and 5.

![Screenshot 2021-04-28 125652](https://user-images.githubusercontent.com/8360474/116385403-6f528c00-a821-11eb-8a7d-472d4174a20a.jpg)

**AUTO_INCREMENT** value should be 4 + 2 = 6, but it is 9.

![Screenshot 2021-04-28 125722](https://user-images.githubusercontent.com/8360474/116385420-74174000-a821-11eb-886d-272f4fb56816.jpg)

3. Let's uncheck for Related Products: 3 and click [Save and Continue Edit]. Table catalog_product_link has the AUTO_INCREMENT value set to 6 which is correct. A strange behavior occurs again in table catalog_product_link_attribute_int. See bellow:

**AUTO_INCREMENT** value should be 6, but from 9 jumped to 13, just unchecking a value!

![Screenshot 2021-04-28 130453](https://user-images.githubusercontent.com/8360474/116386356-5bf3f080-a822-11eb-994f-46b69e737546.jpg)

**Comment by ADDISON74** [comment]
**IMPORTING PRODUCTS - THE WHOLE EXPORTED CSV FILE**

Both **catalog_product_link** and **catalog_product_link_attribute_int** tables have auto_increment values set to 1. There are 5 simple products created before. I am creating a new simple product then add for Related Products: 1, 2, 3 and for Up-sells Product: 4, 5 and click [Save and Continue Edit]. Now in database both auto_increment values are set to 6 which is correct.

Let's export from Backend all information related to Products. Once you get the CSV file then import it from Backend with behavior "Replace Existing Complex Data". See the results in database for both tables.

For **catalog_product_link** what was before is deleted then is replaced by CSV information. **link_id** value starts correctly with 6. The new value is 11 also correct.

![Screenshot 2021-04-28 134552](https://user-images.githubusercontent.com/8360474/116391616-11757280-a828-11eb-9462-fd70c1127c14.jpg)

For **catalog_product_link** what was before is deleted and the table remain empty. The auto_increment value was changed from 6 to 9.

![Screenshot 2021-04-28 134910](https://user-images.githubusercontent.com/8360474/116391977-89439d00-a828-11eb-998f-0153b60c183b.jpg)

In conclusion, the import process is generating this issue. catalog_product_link_attribute_int is important because it has information about position. I visited product 6 and pressed the [Save] button and nothing changed into this table, Magento considers it has values. 

I will create a new scenarios by deleting most of the columns in CSV file keeping only SKU and Price and see what is happening.

**Comment by ADDISON74** [comment]
**IMPORTING PRODUCTS - RELATED / UPSELL / CROSS-SELL WITH POSITION**

Both **catalog_product_link** and **catalog_product_link_attribute_int** tables have auto_increment values set to 1. There are 5 simple products created before. I am creating a new simple product then add for Related Products: 1, 2, 3, for Up-sells Product: 4, 5, and the same values as position for each of them and click [Save and Continue Edit]. 

Let's export from Backend all information related to Products. 

![with_position_set](https://user-images.githubusercontent.com/8360474/116463663-a2be0680-a873-11eb-83a6-704627d55f3b.jpg)

Once you get the CSV file then import it from Backend with behavior "Replace Existing Complex Data". See the results in database for both tables. All data were imported correctly in tables, especially catalog_product_link_attribute_int.

Let's set all values for Position to zero then export again all information related to Products.

![with_position_zero](https://user-images.githubusercontent.com/8360474/116464108-3c85b380-a874-11eb-9689-460f77820f3a.jpg)

Import the CSV file with the same behavior "Replace Existing Complex data". See the results in database for both tables. In table catalog_product_link everything looks fine, but for table catalog_product_link_attribute_int things don't look good at all. Instead of having 5 rows there is an empty table. See bellow:

![empty_table](https://user-images.githubusercontent.com/8360474/116464592-cdf52580-a874-11eb-8293-5d770184b7ca.jpg)

If you go and edit product 6 and click the [Save] button catalog_product_link_attribute_int table remains the same. But if you visit one after other the sections from left Related Products/Up-sells/Cross-sells and click [Save and Continue Edit] ]the table will get records set to zero.

In conclusion, if there is in the imported CSV file a value of zero for Position, it won't be imported. Actually no rows will be inserted into the table. This is a bug in Magento core.

**Comment by ADDISON74** [comment]
**In conclusion**

My personal opinion is there are bugs in OpenMage code.

1. auto_increment for table catalog_product_link_attribute_int has an issue. instead of incrementing 1 by 1 it jumps probably based on the number of links in a section
2. if there is in the imported CSV file a value of zero for Position, it won't be imported. Actually no rows will be inserted into the table.

If someone with good knowledge could check the code for product links it will great.

**Comment by henryhayes** [comment]
> My personal opinion is there are bugs in OpenMage code.

Your opinion is not what we need here. If you cannot point to the exact piece of code where you believe the bug is, then we need you to approach this differently.

Please start with a clean version of OpenMage [you can use [andreaskoch/dockerized-magento]([andreaskoch/dockerized-magento](https://github.com/andreaskoch/dockerized-magento)) or the superchraged [julienloizelet/docker-openmage](https://github.com/julienloizelet/docker-openmage)] and try this from scratch on lts code with no modules installed.

**Comment by ADDISON74** [comment]
@henryhayes - With all due respect when you express yourself please do not use the wording "we need here" or "we need you". After all, you express a personal opinion like me and you do not speak on behalf of a group of people. Thank you for your understanding.

I made a detailed report and left it open. If someone is facing this situation and is smarter than me he will most likely offer me a solution.
